core/arch/instr_shared.c

```diff
@@ -98,10 +98,12 @@ instr_create(dcontext_t *dcontext)
     return instr;
 }
 
-/* deletes the instr_t object with handle "inst" and frees its storage */
+/* deletes the instr_t object with handle "instr" and frees its storage */
 void
 instr_destroy(dcontext_t *dcontext, instr_t *instr)
 {
+    if (instr_is_label(instr) && instr_get_label_callback(instr) != NULL)
```

Comment:
This needs to be in instr_free(), not here.  instr_destroy() is sometimes never called (for stack-allocated instr_t): it's instr_free() that frees the resources inside the instr_t.

Reply:
Ah ok. Moved to `instr_free()` and tested.